src/Microsoft.AspNetCore.ResponseCaching/Internal/DistributedResponseCacheStore.cs

```diff
@@ -42,17 +60,51 @@ public DistributedResponseCacheStore(IDistributedCache cache)
             catch { }
         }
 
-        public async Task SetAsync(string key, object entry, TimeSpan validFor)
+        public async Task SetAsync(string key, IResponseCacheEntry entry, TimeSpan validFor)
         {
             try
             {
+                var cachedResponse = entry as CachedResponse;
+                if (cachedResponse != null)
+                {
+                    cachedResponse.BodyKeyPrefix = FastGuid.NewGuid().IdString;
+                }
+
                 await _cache.SetAsync(
                     key,
-                    CacheEntrySerializer.Serialize(entry),
+                    ResponseCacheEntrySerializer.Serialize(entry),
                     new DistributedCacheEntryOptions()
                     {
                         AbsoluteExpirationRelativeToNow = validFor
                     });
+
+                if (cachedResponse != null)
+                {
+                    var shardCount = (cachedResponse.Body.Length + _options.CachedBodyShardSize - 1) / _options.CachedBodyShardSize;
+
+                    for (int i = 0; i < shardCount; i++)
+                    {
+                        // TODO: doesn't need a new shard every time?
+                        var shard = new byte[_options.CachedBodyShardSize];
+                        var bytesRead = cachedResponse.Body.Read(shard, 0, _options.CachedBodyShardSize);
+
+                        // The last shard may not be full
+                        if (bytesRead != _options.CachedBodyShardSize)
```

Comment:
Can this happen due to the underlying store for any of the other shards (not just the last one)? Perhaps, for example, reads not completing fully. 

Also, could `bytesRead` be ever 0?